modules/catalog-compaction/src/main/java/org/apache/ignite/internal/catalog/compaction/CatalogCompactionRunner.java

```diff
@@ -401,74 +433,35 @@ private static List<String> missingNodes(Set<String> requiredNodes, Collection<L
+    private CompletableFuture<Void> invokeOnLocalReplicas(long txBeginTime, ObjectIterator<Entry> tabTtr) {
+        if (!tabTtr.hasNext()) {
+            return CompletableFutures.nullCompletedFuture();
         }
 
-        return placementDriver.getAssignments(replicationGroupIds, nowTs)
-                .thenComposeAsync(tokenizedAssignments -> {
-                    assert tokenizedAssignments.size() == replicationGroupIds.size();
-
-                    List<CompletableFuture<?>> replicaInvokeFutures = new ArrayList<>(partitions);
-
-                    for (int p = 0; p < partitions; p++) {
-                        TablePartitionId replicationGroupId = replicationGroupIds.get(p);
-                        TokenizedAssignments tokenizedAssignment = tokenizedAssignments.get(p);
+        Entry tableWithPartitions = tabTtr.next();
+        int tableId = tableWithPartitions.getIntKey();
+        int partitions = tableWithPartitions.getIntValue();
+        List<CompletableFuture<?>> partFutures = new ArrayList<>(partitions);
+        HybridTimestamp nowTs = clockService.now();
 
-                        if (tokenizedAssignment == null) {
-                            throwAssignmentsNotReadyException(replicationGroupId);
+        for (int p = 0; p < partitions; p++) {
+            TablePartitionId tablePartitionId = new TablePartitionId(tableId, p);
+
+            CompletableFuture<?> fut = placementDriver
+                    .getPrimaryReplica(tablePartitionId, nowTs)
+                    .thenCompose(meta -> {
+                        // If primary is not elected yet - we'll update replication groups on next iteration.
+                        if (meta == null || meta.getLeaseholder() == null) {
+                            return CompletableFutures.nullCompletedFuture();
                         }
 
-                        Set<String> assignments = tokenizedAssignment.nodes().stream()
-                                .map(Assignment::consistentId).collect(Collectors.toSet());
-
-                        String targetNodeName;
-
-                        if (assignments.contains(localNodeName)) {
-                            targetNodeName = localNodeName;
-                        } else {
-                            targetNodeName = assignments.stream()
-                                    .filter(logicalTopologyNodes::contains)
-                                    .findAny()
-                                    .orElseThrow(() -> new IllegalStateException("Current topology doesn't include assignment nodes "
-                                            + "(assignments=" + tokenizedAssignment.nodes()
-                                            + ", topology=" + logicalTopologyNodes
-                                            + ", replication group=" + replicationGroupId + ").")
-                                    );
+                        if (!localNodeName.equals(meta.getLeaseholder())) {
```

Comment:
Please check nodeId for equality instead of a node name here.

Reply:
Fixed, thanks.